File: bindings/rust/src/chart.rs
//! chart module for Rust bindings.
use crate::ffi;
use crate::types::*;

/// Point & Figure chart for technical analysis.
pub struct Chart {
    ptr: *mut ffi::PnfChart,
}

impl Chart {
    /// Creates a new chart with default configuration.
    ///
    /// # Returns
    /// 
    pub fn new() -> Self {
        let ptr = unsafe { ffi::pnf_chart_create_default() };
        Self { ptr }
    }

    /// Creates a new chart with the specified configuration.
    ///
    /// # Arguments
    /// * `config` - 
    ///
    /// # Returns
    /// 
    pub fn with_config(config: ChartConfig) -> Self {
        let ffi_config = config.to_ffi();
        let ptr = unsafe { ffi::pnf_chart_create(&ffi_config) };
        Self { ptr }
    }

    /// Adds OHLC data to the chart.
    ///
    /// # Arguments
    /// * `&mut self` - 
    /// * `high` - 
    /// * `low` - 
    /// * `close` - 
    /// * `timestamp` - 
    ///
    /// # Returns
    /// 
    pub fn add_data(&mut self, high: f64, low: f64, close: f64, timestamp: i64) -> bool {
        unsafe { ffi::pnf_chart_add_data(self.ptr, high, low, close, timestamp) }
    }

    /// Adds a single price point to the chart.
    ///
    /// # Arguments
    /// * `&mut self` - 
    /// * `price` - 
    /// * `timestamp` - 
    ///
    /// # Returns
    /// 
    pub fn add_price(&mut self, price: f64, timestamp: i64) -> bool {
        unsafe { ffi::pnf_chart_add_price(self.ptr, price, timestamp) }
    }

    /// Adds OHLC data using the OHLC struct.
    ///
    /// # Arguments
    /// * `&mut self` - 
    /// * `ohlc` - 
    ///
    /// # Returns
    /// 
    pub fn add_ohlc(&mut self, ohlc: &OHLC) -> bool {
        let ffi_ohlc = ohlc.to_ffi();
        unsafe { ffi::pnf_chart_add_ohlc(self.ptr, &ffi_ohlc) }
    }

    /// Returns the total number of columns in the chart.
    ///
    /// # Arguments
    /// * `&self` - 
    ///
    /// # Returns
    /// 
    pub fn column_count(&self) -> usize {
        unsafe { ffi::pnf_chart_column_count(self.ptr) }
    }

    /// Returns the number of X columns.
    ///
    /// # Arguments
    /// * `&self` - 
    ///
    /// # Returns
    /// 
    pub fn x_column_count(&self) -> usize {
        unsafe { ffi::pnf_chart_x_column_count(self.ptr) }
    }

    /// Returns the number of O columns.
    ///
    /// # Arguments
    /// * `&self` - 
    ///
    /// # Returns
    /// 
    pub fn o_column_count(&self) -> usize {
        unsafe { ffi::pnf_chart_o_column_count(self.ptr) }
    }

    /// Returns the current box size.
    ///
    /// # Arguments
    /// * `&self` - 
    ///
    /// # Returns
    /// 
    pub fn box_size(&self) -> f64 {
        unsafe { ffi::pnf_chart_box_size(self.ptr) }
    }

    /// Returns the type of column at the specified index.
    ///
    /// # Arguments
    /// * `&self` - 
    /// * `index` - 
    ///
    /// # Returns
    /// 
    pub fn column_type(&self, index: usize) -> ColumnType {
        let ct = unsafe { ffi::pnf_chart_column_type(self.ptr, index) };
        ColumnType::from_ffi(ct)
    }

    /// Returns the number of boxes in the specified column.
    ///
    /// # Arguments
    /// * `&self` - 
    /// * `index` - 
    ///
    /// # Returns
    /// 
    pub fn column_box_count(&self, index: usize) -> usize {
        unsafe { ffi::pnf_chart_column_box_count(self.ptr, index) }
    }

    /// Returns the highest price in the specified column.
    ///
    /// # Arguments
    /// * `&self` - 
    /// * `index` - 
    ///
    /// # Returns
    /// 
    pub fn column_highest(&self, index: usize) -> f64 {
        unsafe { ffi::pnf_chart_column_highest(self.ptr, index) }
    }

    /// Returns the lowest price in the specified column.
    ///
    /// # Arguments
    /// * `&self` - 
    /// * `index` - 
    ///
    /// # Returns
    /// 
    pub fn column_lowest(&self, index: usize) -> f64 {
        unsafe { ffi::pnf_chart_column_lowest(self.ptr, index) }
    }

    /// Returns the price of a box at the specified column and box index.
    ///
    /// # Arguments
    /// * `&self` - 
    /// * `col_index` - 
    /// * `box_index` - 
    ///
    /// # Returns
    /// 
    pub fn box_price(&self, col_index: usize, box_index: usize) -> f64 {
        unsafe { ffi::pnf_chart_box_price(self.ptr, col_index, box_index) }
    }

    /// Returns the type of box at the specified column and box index.
    ///
    /// # Arguments
    /// * `&self` - 
    /// * `col_index` - 
    /// * `box_index` - 
    ///
    /// # Returns
    /// 
    pub fn box_type(&self, col_index: usize, box_index: usize) -> BoxType {
        let bt = unsafe { ffi::pnf_chart_box_type(self.ptr, col_index, box_index) };
        BoxType::from_ffi(bt)
    }

    /// Returns the marker string for a box at the specified position.
    ///
    /// # Arguments
    /// * `&self` - 
    /// * `col_index` - 
    /// * `box_index` - 
    ///
    /// # Returns
    /// 
    pub fn box_marker(&self, col_index: usize, box_index: usize) -> Option<String> {
        let ptr = unsafe { ffi::pnf_chart_box_marker(self.ptr, col_index, box_index) };
        let result = cstr_to_string(ptr);
        if !ptr.is_null() {
            unsafe { ffi::pnf_free_string(ptr) };
        }
        result
    }

    /// Checks if the chart has a bullish bias.
    ///
    /// # Arguments
    /// * `&self` - 
    ///
    /// # Returns
    /// 
    pub fn has_bullish_bias(&self) -> bool {
        unsafe { ffi::pnf_chart_has_bullish_bias(self.ptr) }
    }

    /// Checks if the chart has a bearish bias.
    ///
    /// # Arguments
    /// * `&self` - 
    ///
    /// # Returns
    /// 
    pub fn has_bearish_bias(&self) -> bool {
        unsafe { ffi::pnf_chart_has_bearish_bias(self.ptr) }
    }

    /// Checks if the price is above bullish support line.
    ///
    /// # Arguments
    /// * `&self` - 
    /// * `price` - 
    ///
    /// # Returns
    /// 
    pub fn is_above_bullish_support(&self, price: f64) -> bool {
        unsafe { ffi::pnf_chart_is_above_bullish_support(self.ptr, price) }
    }

    /// Checks if the price is below bearish resistance line.
    ///
    /// # Arguments
    /// * `&self` - 
    /// * `price` - 
    ///
    /// # Returns
    /// 
    pub fn is_below_bearish_resistance(&self, price: f64) -> bool {
        unsafe { ffi::pnf_chart_is_below_bearish_resistance(self.ptr, price) }
    }

    /// Clears all data from the chart.
    ///
    /// # Arguments
    /// * `&mut self` - 
    pub fn clear(&mut self) {
        unsafe { ffi::pnf_chart_clear(self.ptr) }
    }

    /// Returns a string representation of the chart.
    ///
    /// # Arguments
    /// * `&self` - 
    ///
    /// # Returns
    /// 
    pub fn to_string(&self) -> Option<String> {
        let ptr = unsafe { ffi::pnf_chart_to_string(self.ptr) };
        let result = cstr_to_string(ptr);
        if !ptr.is_null() {
            unsafe { ffi::pnf_free_string(ptr) };
        }
        result
    }

    /// Returns an ASCII representation of the chart.
    ///
    /// # Arguments
    /// * `&self` - 
    ///
    /// # Returns
    /// 
    pub fn to_ascii(&self) -> Option<String> {
        let ptr = unsafe { ffi::pnf_chart_to_ascii(self.ptr) };
        let result = cstr_to_string(ptr);
        if !ptr.is_null() {
            unsafe { ffi::pnf_free_string(ptr) };
        }
        result
    }

    /// Returns a JSON representation of the chart.
    ///
    /// # Arguments
    /// * `&self` - 
    ///
    /// # Returns
    /// 
    pub fn to_json(&self) -> Option<String> {
        let ptr = unsafe { ffi::pnf_chart_to_json(self.ptr) };
        let result = cstr_to_string(ptr);
        if !ptr.is_null() {
            unsafe { ffi::pnf_free_string(ptr) };
        }
        result
    }

    pub(crate) fn as_ptr(&self) -> *const ffi::PnfChart {
        self.ptr as *const ffi::PnfChart
    }
}

impl Default for Chart {
    fn default() -> Self {
        Self::new()
    }
}

impl Drop for Chart {
    fn drop(&mut self) {
        unsafe {
            ffi::pnf_chart_destroy(self.ptr);
        }
    }
}

unsafe impl Send for Chart {}
unsafe impl Sync for Chart {}

File: bindings/csharp/Chart.cs
using System;
using System.Runtime.InteropServices;

namespace PnF
{
    /// <summary>
    /// Managed wrapper around a native Point and Figure chart.
    /// </summary>
    public class Chart : IDisposable
    {
        private IntPtr _handle;
        private bool _disposed;

        public Chart()
        {
            _handle = NativeMethods.pnf_chart_create_default();
            if (_handle == IntPtr.Zero)
                throw new InvalidOperationException("Failed to create chart");
        }

        public Chart(ChartConfig config)
        {
            _handle = NativeMethods.pnf_chart_create(ref config);
            if (_handle == IntPtr.Zero)
                throw new InvalidOperationException("Failed to create chart with config");
        }

        /// <summary>
        /// AddData.
        /// </summary>
        /// <param name="high"></param>
        /// <param name="low"></param>
        /// <param name="close"></param>
        /// <param name="timestamp"></param>
        /// <returns></returns>
        public bool AddData(double high, double low, double close, long timestamp)
        {
            ThrowIfDisposed();
            return NativeMethods.pnf_chart_add_data(_handle, high, low, close, timestamp);
        }

        /// <summary>
        /// AddPrice.
        /// </summary>
        /// <param name="price"></param>
        /// <param name="timestamp"></param>
        /// <returns></returns>
        public bool AddPrice(double price, long timestamp)
        {
            ThrowIfDisposed();
            return NativeMethods.pnf_chart_add_price(_handle, price, timestamp);
        }

        /// <summary>
        /// AddOHLC.
        /// </summary>
        /// <param name="ohlc"></param>
        /// <returns></returns>
        public bool AddOHLC(OHLC ohlc)
        {
            ThrowIfDisposed();
            return NativeMethods.pnf_chart_add_ohlc(_handle, ref ohlc);
        }

        public int ColumnCount
        {
            get
            {
                ThrowIfDisposed();
                return (int)NativeMethods.pnf_chart_column_count(_handle);
            }
        }

        public int XColumnCount
        {
            get
            {
                ThrowIfDisposed();
                return (int)NativeMethods.pnf_chart_x_column_count(_handle);
            }
        }

        public int OColumnCount
        {
            get
            {
                ThrowIfDisposed();
                return (int)NativeMethods.pnf_chart_o_column_count(_handle);
            }
        }

        public double BoxSize
        {
            get
            {
                ThrowIfDisposed();
                return NativeMethods.pnf_chart_box_size(_handle);
            }
        }

        /// <summary>
        /// GetColumnType.
        /// </summary>
        /// <param name="index"></param>
        /// <returns></returns>
        public ColumnType GetColumnType(int index)
        {
            ThrowIfDisposed();
            return NativeMethods.pnf_chart_column_type(_handle, (UIntPtr)index);
        }

        /// <summary>
        /// GetColumnBoxCount.
        /// </summary>
        /// <param name="index"></param>
        /// <returns></returns>
        public int GetColumnBoxCount(int index)
        {
            ThrowIfDisposed();
            return (int)NativeMethods.pnf_chart_column_box_count(_handle, (UIntPtr)index);
        }

        /// <summary>
        /// GetColumnHighest.
        /// </summary>
        /// <param name="index"></param>
        /// <returns></returns>
        public double GetColumnHighest(int index)
        {
            ThrowIfDisposed();
            return NativeMethods.pnf_chart_column_highest(_handle, (UIntPtr)index);
        }

        /// <summary>
        /// GetColumnLowest.
        /// </summary>
        /// <param name="index"></param>
        /// <returns></returns>
        public double GetColumnLowest(int index)
        {
            ThrowIfDisposed();
            return NativeMethods.pnf_chart_column_lowest(_handle, (UIntPtr)index);
        }

        /// <summary>
        /// GetBoxPrice.
        /// </summary>
        /// <param name="columnIndex"></param>
        /// <param name="boxIndex"></param>
        /// <returns></returns>
        public double GetBoxPrice(int columnIndex, int boxIndex)
        {
            ThrowIfDisposed();
            return NativeMethods.pnf_chart_box_price(_handle, (UIntPtr)columnIndex, (UIntPtr)boxIndex);
        }

        /// <summary>
        /// GetBoxType.
        /// </summary>
        /// <param name="columnIndex"></param>
        /// <param name="boxIndex"></param>
        /// <returns></returns>
        public BoxType GetBoxType(int columnIndex, int boxIndex)
        {
            ThrowIfDisposed();
            return NativeMethods.pnf_chart_box_type(_handle, (UIntPtr)columnIndex, (UIntPtr)boxIndex);
        }

        public string? GetBoxMarker(int columnIndex, int boxIndex)
        {
            ThrowIfDisposed();
            var ptr = NativeMethods.pnf_chart_box_marker(_handle, (UIntPtr)columnIndex, (UIntPtr)boxIndex);
            if (ptr == IntPtr.Zero) return null;
            var result = Marshal.PtrToStringAnsi(ptr);
            NativeMethods.pnf_free_string(ptr);
            return result;
        }

        public bool HasBullishBias
        {
            get
            {
                ThrowIfDisposed();
                return NativeMethods.pnf_chart_has_bullish_bias(_handle);
            }
        }

        public bool HasBearishBias
        {
            get
            {
                ThrowIfDisposed();
                return NativeMethods.pnf_chart_has_bearish_bias(_handle);
            }
        }

        /// <summary>
        /// IsAboveBullishSupport.
        /// </summary>
        /// <param name="price"></param>
        /// <returns></returns>
        public bool IsAboveBullishSupport(double price)
        {
            ThrowIfDisposed();
            return NativeMethods.pnf_chart_is_above_bullish_support(_handle, price);
        }

        /// <summary>
        /// IsBelowBearishResistance.
        /// </summary>
        /// <param name="price"></param>
        /// <returns></returns>
        public bool IsBelowBearishResistance(double price)
        {
            ThrowIfDisposed();
            return NativeMethods.pnf_chart_is_below_bearish_resistance(_handle, price);
        }

        /// <summary>
        /// Clear.
        /// </summary>
        public void Clear()
        {
            ThrowIfDisposed();
            NativeMethods.pnf_chart_clear(_handle);
        }

        /// <summary>
        /// Returns a readable summary string.
        /// </summary>
        /// <returns></returns>
        public override string ToString()
        {
            ThrowIfDisposed();
            var ptr = NativeMethods.pnf_chart_to_string(_handle);
            if (ptr == IntPtr.Zero) return string.Empty;
            var result = Marshal.PtrToStringAnsi(ptr) ?? string.Empty;
            NativeMethods.pnf_free_string(ptr);
            return result;
        }

        /// <summary>
        /// ToAscii.
        /// </summary>
        /// <returns></returns>
        public string ToAscii()
        {
            ThrowIfDisposed();
            var ptr = NativeMethods.pnf_chart_to_ascii(_handle);
            if (ptr == IntPtr.Zero) return string.Empty;
            var result = Marshal.PtrToStringAnsi(ptr) ?? string.Empty;
            NativeMethods.pnf_free_string(ptr);
            return result;
        }

        /// <summary>
        /// ToJson.
        /// </summary>
        /// <returns></returns>
        public string ToJson()
        {
            ThrowIfDisposed();
            var ptr = NativeMethods.pnf_chart_to_json(_handle);
            if (ptr == IntPtr.Zero) return string.Empty;
            var result = Marshal.PtrToStringAnsi(ptr) ?? string.Empty;
            NativeMethods.pnf_free_string(ptr);
            return result;
        }

        internal IntPtr Handle => _handle;

        private void ThrowIfDisposed()
        {
            if (_disposed)
                throw new ObjectDisposedException(nameof(Chart));
        }

        protected virtual void Dispose(bool disposing)
        {
            if (!_disposed)
            {
                if (_handle != IntPtr.Zero)
                {
                    NativeMethods.pnf_chart_destroy(_handle);
                    _handle = IntPtr.Zero;
                }
                _disposed = true;
            }
        }

        ~Chart()
        {
            Dispose(false);
        }

        /// <summary>
        /// Releases unmanaged resources.
        /// </summary>
        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }
    }
}
